src/react_native/core.cljs

```diff
@@ -135,25 +135,34 @@
 (def use-context react/useContext)
 
 (defn use-effect
-  ([effect-fn]
-   (use-effect effect-fn []))
-  ([effect-fn deps]
+  {:deprecated
+   "use-mount or use-unmount should be used, more here https://github.com/status-im/status-mobile/blob/develop/doc/ui-guidelines.md#effects"}
+  ([handler]
+   (use-effect handler nil))
+  ([handler deps]
    (react/useEffect
-    #(let [ret (effect-fn)]
+    #(let [ret (handler)]
        (if (fn? ret) ret js/undefined))
-    (bean/->js deps))))
+    (if deps (bean/->js deps) js/undefined))))
```

Comment:
Since we're converting the deps to js structures, should we be careful with cases where the dependencies are maps/vectors i.e. converted into reference types (objects/arrays). Would `bean/->js` return the same object reference if the object/array didn't change? Usually with JS we'd have to wrap them in a `useMemo` to make sure it's the same.

Reply:
very good point, most likely bean/->js shouldn't be used here, we should compare cljs data, same for use-callback, so currently we have only primitive data types, that's why it works

Reply:
hm actually that's interesting, how it works, most likely rect stores it somewhere, and then compares with `=` so basically pure cljs data should work?

Reply:
for arrays/objects it would end up comparing the references instead of values i think

Reply:
`(to-array deps)` should work better i guess

Reply:
i need to experiment with that

Reply:
UIx doesn't seem to be [doing anything about it](https://pitch-io.github.io/uix/docs/hooks.html) , just suggesting to use primitive values only. Wonder if smth like this would work.

```clj
(defn use-effect
  [handler deps]
  (let [prev-deps     (use-ref-atom deps)
        deps-changed? (not (every? = @prev-deps deps))]
    (when deps-changed?
      (reset! prev-deps deps))

    (use-effect handler @prev-deps)))

```

Reply:
@clauxx yes you're right it won't work, for memo there is optional arePropsEqual: A function that accepts two arguments, which can be used to compare cljs data, but im not sure what we can do for useEffect and useCallback any ideas? im not sure we ever need to compare maps or arrays, but still

Reply:
probably for now we could have a check and fire an error in dev builds ? wdyt

Reply:
thanks for the uix link, i will take a look

Reply:
I tried a simple example with doing the comparison on the cljs side and it seems to work. At this stage though just validating (e.g. with malli) against primitive values in dev builds might be wiser.

``` clj
;; could be replaced with the use-ref-atom
(def prev-deps
  (atom {:deps    []
         :js-deps (bean/->js [])}))

(defn use-effect
  [handler deps]
  (let [deps-changed? (not= deps (:deps @prev-deps))]
    (println @prev-deps)
    (when deps-changed?
      (reset! prev-deps {:deps    deps
                         :js-deps (bean/->js deps)}))
    (rn/use-effect handler (:js-deps @prev-deps))))
```

Reply:
@flexsurfer, @clauxx The deps are compared with `Object.is`. I would personally prefer if we used the principle of least surprise and respected the semantics of this function. React devs know the deps should not contain objects or arrays.

If we start to encourage the usage of arrays and objects in deps by supporting this use case, it will just lead to more bad examples not respecting the guidelines, such as the one you wrote about the _lifecycle_ of effects.

> probably for now we could have a validation and fire an error in dev builds ? wdyt

If I understand your suggestion, we can just instrument `use-effect` with malli.

Reply:
at the very least if an object/array has to be used as a dependency, we can advise to wrap them with a `use-memo`. Seems to be pretty common with react (but in react 19 as far as I understood useMemo/useCallback won't be needed anymore).

Reply:
agree with you, we could add malli and upgrade guidelines later, but still i preffer to keep compare fn on our side

Reply:
+1 
I will say that for me it's a bit strange to be adding new code and guidelines at the same time.
Imo it's best to leave this idea for now, make a follow up just adding that code, test it out. make sure it works right and then provide best practices for use-effect using objects/arrays as dependencies.
For now just don't handle it or throw an error etc 👍